### PR TITLE
Automated wheel building

### PR DESCRIPTION
**Pull request by mmulich**
This builds the wheel for linux and macos as well as the source distribution. It then publishes the results to PyPI.

### CODE REVIEW

**Review by bluemellophone — CHANGES_REQUESTED** [review]
I would like to standardize all our builds using cibuildwheel, here is an following example: 
https://github.com/joerick/cibuildwheel/blob/master/examples/github-deploy.yml

File: vtool/__init__.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 """
 # flake8: noqa
 from __future__ import absolute_import, division, print_function
-__version__ = '2.1.1'
+__version__ = '2.0.0'
```
Comment (karenc) [inline vtool/__init__.py]: There's already a `v2.0.0` tag so I think it might be better to go with `v3.0.0`?  Or should we go with `v2.2.0`?

Comment (mmulich) [reply vtool/__init__.py]: Changed to 3.0.0 👍

**Comment by bluemellophone** [comment]
> This builds the wheel for linux and macos as well as the source distribution. It then publishes the results to PyPI.
> 
> # DO NOT MERGE
> This is currently using `mmulich/manylinux` because the docker hub perms for our bot aren't working. More info in the group chat... I'll update this PR once we've got that resolved. You can see the work around for this in the 💀 commit. Bug for this filed at https://wildme.atlassian.net/browse/WB-561

Is this PR still under the restriction of `DO NOT MERGE`?

File: vtool/__init__.py [hunk 2]
```diff
@@ -7,7 +7,7 @@
 """
 # flake8: noqa
 from __future__ import absolute_import, division, print_function
-__version__ = '2.1.1'
+__version__ = '3.0.0'
```
Comment (bluemellophone) [inline vtool/__init__.py:10]: I approve major version bump for wbia-vtool

File: .github/workflows/python-publish.yml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      - name: Build wheels
+        env:
+          # FIXME (19-Jun-12020) Having permissions issues with docker hub
+          # CIBW_MANYLINUX_X86_64_IMAGE: wildme/manylinux:latest
+          CIBW_MANYLINUX_X86_64_IMAGE: mmulich/manylinux:latest
+          CIBW_BUILD: cp36-manylinux_x86_64 cp36-macosx_x86_64
+          # Script to prepare the macos environment
+          CIBW_BEFORE_BUILD_MACOS: bash scripts/ci_prepare_macos_for_build.sh
+        run: |
+          python -m cibuildwheel --output-dir wheelhouse
```
Comment (bluemellophone) [inline .github/workflows/python-publish.yml:48]: Is there a reason we cannot use `cibuildwheel` on Ubuntu as well?

File: .github/workflows/python-publish.yml [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        name: Install Python
+        with:
+          python-version: '3.6'
+
+      - name: Install cibuildwheel
+        run: |
+          python -m pip install cibuildwheel==1.4.2
+
+      - name: Cache OpenCV dependency for macOS
+        if: ${{ runner.os == 'macOS' }}
```
Comment (bluemellophone) [inline .github/workflows/python-publish.yml:28]: How does OpenCV get built and cached in the Ubuntu environment?

File: CMakeLists.txt [hunk 5]
```diff
@@ -35,13 +35,15 @@ message(STATUS "VTOOL_VERSION = ${VTOOL_VERSION}")
 DISSECT_VERSION()
 GET_OS_INFO()
 
-if (OS_IS_MACOS)
-    message(STATUS "INCLUDING MACPORTS")
+if (APPLE)
+    message(STATUS "Including Homebrew")
```
Comment (bluemellophone) [inline CMakeLists.txt:39]: I'm moderately opposed to moving our build environment on macOS to Homebrew over MacPorts.  However, I will concede that using MacPorts is a big issue with CI since it requires Xcode to be installed and configured correctly prior to running.  It also helps that GitHub Actions' macOS VM supports Homebrew as the package manager of choice (https://github.com/actions/virtual-environments/blob/master/images/macos/macos-10.15-Readme.md).  I find MacPorts to be updated more frequently and has better support for compilers.

Luckily, it looks like Xcode is installed already in that environment and installing MacPorts seems to be pretty much supported already OOTB with the help of this GitHub action in the marketplace: https://github.com/keepitsimpletech/macports-github-actions/blob/master/.github/workflows/macports.yml

File: scripts/_install_opencv_build_dependencies_on_macos.sh [hunk 6]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+################################
+# Install OpenCV dependencies
+
+set -ex
+
+brew install pkg-config
```
Comment (bluemellophone) [inline scripts/_install_opencv_build_dependencies_on_macos.sh:7]: If we pivot to MacPorts, I can provide the equivalent for these libraries.

Comment (bluemellophone) [reply scripts/_install_opencv_build_dependencies_on_macos.sh:7]: Currently, I install these MacPorts on macOS for OpenCV building against MacPorts' Clang and LLVM 5.0 with OpenMP support:

```
sudo port install lapack eigen3-devel OpenBLAS-devel hdf5 py36-h5py-devel gdal py36-gdal freetype tesseract harfbuzz google-glog leptonica clang-5.0 llvm-5.0 libomp clang_select llvm_select

sudo port select clang mp-clang-5.0
sudo port select llvm mp-llvm-5.0
```

**Review by bluemellophone — CHANGES_REQUESTED** [review]
I'd like to revisit switching to Homebrew over MacPorts.  It will require us to upgrade a bunch of MacPort library mappings down the line and I'm not convinced I want to personally move away from MacPorts for personal development.  I'd prefer my development environment on macOS mirror the CI build and testing process we create.

File: .github/workflows/python-publish.yml (hunk 3 above)
Comment (mmulich) [reply .github/workflows/python-publish.yml:48]: It is being used on ubuntu. You don't see it directly, but a few lines up is the matrix definition

File: .github/workflows/python-publish.yml (hunk 4 above)
Comment (mmulich) [reply .github/workflows/python-publish.yml:28]: There is no need for a cache in ubuntu because we have it built into the container image.

File: .github/workflows/python-publish.yml (hunk 3 above)
Comment (bluemellophone) [reply .github/workflows/python-publish.yml:48]: Gotcha, I find these YAML files a bit hard to follow

File: .github/workflows/python-publish.yml (hunk 4 above)
Comment (bluemellophone) [reply .github/workflows/python-publish.yml:28]: 👍

Comment (bluemellophone) [reply .github/workflows/python-publish.yml:28]: Yep, makes sense, I see `wildme/manylinux:latest` being used as the build context for `cibuildwheel`

File: CMakeLists.txt (hunk 5 above)
Comment (mmulich) [reply CMakeLists.txt:39]: I was hoping to quickly throw together a solution based on this info. My hopes have been smashed... 

https://github.com/keepitsimpletech/actions/tree/master/setup-macports is now 10 months from its last update. It was originally built for macos 10.14. I'd think the action would have received an update if the author were still using this action, because their builds would be failing.

Because it is built for macos 10.14 and not receiving updates, we'll likely need to take ownership of the action code if we intend to use it. ~You up for~ _Are you onboard with wildme_ taking ownership of this code?

---

The failure currently looks like this:
```
Run keepitsimpletech/actions/setup-macports@setup-macports/releases/v1
  with:
    UseHostCMake: true
  env:
    pythonLocation: /Users/runner/hostedtoolcache/Python/3.6.10/x64
sudo /usr/sbin/installer -pkg /Users/runner/temp/temp_264321729/MacPorts-2.5.4-10.14-Mojave.pkg -target /
installer: Error - This package is meant to be installed on Mac OS X 10.14.
##[error]The process 'sudo' failed with exit code 1
```

It may be as easy as fixing the download location and version number in the TypeScript code. Difficult to say if that's the only change that would be necessary.

Comment (mmulich) [reply CMakeLists.txt:39]: In the meantime I'll look for other options...

What do you think about *including both macports and homebrew support*? I don't think there is any harm in including both, for example:
```
include_directories(/opt/local/include)
link_directories(/opt/local/lib)
include_directories(/usr/local/include)
link_directories(/usr/local/lib)
```

Then we get the best of both worlds while not hindering development.

Comment (bluemellophone) [reply CMakeLists.txt:39]: Can we build our own MacPorts target similar to this: https://github.com/keepitsimpletech/actions/compare/master...medengineer:master

Comment (bluemellophone) [reply CMakeLists.txt:39]: I'd like to burn a bit longer on the macports support for the immediate term today.  I'm not opposed to us maintaining an updated fork of the code provided by `keepitsimpletech/actions`

Comment (mmulich) [reply CMakeLists.txt:39]: Ok, we'll continue with it. Based on that comparison you showed, it looks like they may have simply moved development to the fork. I had glanced at the fork, but didn't see the adjustment.

I'll contact the author (`medengineer`) about 1) tagging the project 2) outputting the macports version for caching purposes 3) and offering caching instructions in return.

Comment (bluemellophone) [reply CMakeLists.txt:39]: @mmulich I've added an updated MacPorts action to `wbia-pypkg-build` that can be seen here: https://github.com/WildbookOrg/wbia-pypkg-build/actions/runs/146632922.  It creates a `macports` artifact that can be downloaded.  It's linked to a file `macports.tar.gz` that is a tar bundle of the entire `/opt/local/` path on the build machine.  It uses the latest versions of MacPorts for Catalina.

From here, we can start adding dependencies for MacPorts and build them into a bundle.  On deploy, we can download the bundle, extract it into `/opt/local/` on the `pypkg` build instance to get MacPorts configured quickly.  We don't upgrade dependencies in MacPorts often, similar to how our Ubuntu `apt` dependencies are relied on.  We can tag this artifact as a release if it'll make transfer to other repos easy.

Thoughts?

Comment (mmulich) [reply CMakeLists.txt:39]: That could work. Though it's not as aesthetically pleasing as calling action code. You could probably see it as the difference between calling a function on a line vs doing the logic of the function on a line.

Either way, one would likely want to cache the results of the directory. Either direction will work for me, but I'd somewhat favor the action (i.e. medengineer's macports-setup) for clarity and brevity.

Comment (bluemellophone) [reply CMakeLists.txt:39]: I agree, this provides two options for getting MacPorts up and running.  Here are the options I can see: 
1. have the `wbia-pypkg-build` repo build an artifact to load `/opt/local/` onto the build instance as a pre-computed cache
2. Use the action provieded by ` WildbookOrg/wbia-pypkg-build/actions/macports` to get MacPorts installed on the build environment.  This is actually working here (https://github.com/WildbookOrg/wbia-vtool/blob/automated-wheel-w-macports/.github/workflows/python-publish.yml#L37) on the `automated-wheel-w-macports` branch, which is building OpenCV without issue against MacPorts... but is failing to build vtool for some reason.

**Comment by mmulich** [comment]
Closed in favor of a new PR which will include the use of macports.
